pkg/volume/iscsi/iscsi.go

```diff
@@ -0,0 +1,168 @@
+	return ISCSIPluginName
+}
+
+func (plugin *ISCSIPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.ISCSI == nil {
+		return false
+	}
+	// see if iscsiadm is there
+	_, err := plugin.execCommand("iscsiadm", []string{"-h"})
+	if err == nil {
+		return true
+	}
+
+	return false
+}
+
+func (plugin *ISCSIPlugin) GetAccessModes() []api.AccessModeType {
+	return []api.AccessModeType{
+		api.ReadWriteOnce,
+		api.ReadOnlyMany,
+	}
+}
+
+func (plugin *ISCSIPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newBuilderInternal(spec, podRef.UID, &ISCSIUtil{}, mount.New())
+}
+
+func (plugin *ISCSIPlugin) newBuilderInternal(spec *api.Volume, podUID types.UID, manager diskManager, mounter mount.Interface) (volume.Builder, error) {
+	lun := strconv.Itoa(spec.ISCSI.Lun)
+
+	return &iscsiDisk{
+		podUID:   podUID,
+		volName:  spec.Name,
+		portal:   spec.ISCSI.TargetPortal,
+		iqn:      spec.ISCSI.IQN,
+		lun:      lun,
+		fsType:   spec.ISCSI.FSType,
+		readOnly: spec.ISCSI.ReadOnly,
+		manager:  manager,
+		mounter:  mounter,
+		plugin:   plugin,
+	}, nil
+}
+
+func (plugin *ISCSIPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newCleanerInternal(volName, podUID, &ISCSIUtil{}, mount.New())
+}
+
+func (plugin *ISCSIPlugin) newCleanerInternal(volName string, podUID types.UID, manager diskManager, mounter mount.Interface) (volume.Cleaner, error) {
+	return &iscsiDisk{
+		podUID:  podUID,
+		volName: volName,
+		manager: manager,
+		mounter: mounter,
+		plugin:  plugin,
+	}, nil
+}
+
+type iscsiDisk struct {
+	volName  string
+	podUID   types.UID
+	portal   string
+	iqn      string
+	readOnly bool
+	lun      string
+	fsType   string
+	plugin   *ISCSIPlugin
+	mounter  mount.Interface
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager diskManager
+}
+
+func (iscsi *iscsiDisk) GetPath() string {
+	name := ISCSIPluginName
+	// safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
+	return iscsi.plugin.host.GetPodVolumeDir(iscsi.podUID, util.EscapeQualifiedNameForDisk(name), iscsi.volName)
+}
+
+func (iscsi *iscsiDisk) SetUp() error {
+	return iscsi.SetUpAt(iscsi.GetPath())
+}
+
+func (iscsi *iscsiDisk) SetUpAt(dir string) error {
+	// diskSetUp checks mountpoints and prevent repeated calls
+	err := diskSetUp(iscsi.manager, *iscsi, dir, iscsi.mounter)
+	if err != nil {
+		glog.Errorf("iscsi: failed to setup")
+		return err
+	}
+	globalPDPath := iscsi.manager.MakeGlobalPDName(*iscsi)
+	// make mountpoint rw/ro work as expected
+	//FIXME revisit pkg/util/mount and ensure rw/ro is implemented as expected
+	mode := "rw"
+	if iscsi.readOnly {
+		mode = "ro"
+	}
+	iscsi.plugin.execCommand("mount", []string{"-o", "remount," + mode, globalPDPath, dir})
```

Comment:
Why is the generic code using syscall.Mount() instead of doing it like this?  It means that the generic mounter function can never work for things like NFSv3 (binary mount options), gluster, ceph, (fuse), cifs (client server) etc.

mount(8) exists as a helper which does a lot more than just call mount(2)....